Elements shadow dom fix

robwormald opened the pull request:
Adds a new optional parameter to the renderer.selectRootElement method:
Before:

```
renderer.selectRootElement(rootSelectorOrNode);
```
```
renderer.selectRootElement(rootSelectorOrNode, preserveContents?:boolean);
```
`preserveContents` defaults to `false` to preserve existing renderer behavior. If set to true, the renderer will not remove the existing contents of a root element when bootstrapping a component. 

This is primarily useful when combined with the `ViewEncapsulation.ShadowDom` Renderer - the nodes are preserved, and developers can use `<slot>` elements to re-project the light DOM nodes.

This change had to be made to the Renderer API itself, rather than an in a specific renderer as the method is called on the *default* renderer, inside the view engine, rather than a renderer specific to the component: https://github.com/angular/angular/blob/master/packages/core/src/view/services.ts#L756
 
Fixes #24859

robwormald committed:
fix(core): do not clear element content when using shadow dom

mary-poppins commented:
You can preview 1eec884 at https://pr24861-1eec884.ngbuilds.io/.

mary-poppins commented:
You can preview 458cf2f at https://pr24861-458cf2f.ngbuilds.io/.

mhevery commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -135,13 +135,16 @@ class DefaultDomRenderer2 implements Renderer2 {
     }
-    el.textContent = '';
+    if (preserveContent) {
+      return el;
```
Not a fan of early returns. Makes the code hard to follow. Rewrite as:

```
if (!preserveContent) {
  el.textContent = '';
}
```

mhevery commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -271,6 +274,20 @@ class ShadowDomRenderer extends DefaultDomRenderer2 {
     }
   }
 
+  selectRootElement(selectorOrNode: string|any): any {
```
Why is there an extra method here?

mary-poppins commented:
You can preview 525ca3e at https://pr24861-525ca3e.ngbuilds.io/.

mary-poppins commented:
You can preview 7a08d69 at https://pr24861-7a08d69.ngbuilds.io/.

mary-poppins commented:
You can preview 8617d87 at https://pr24861-8617d87.ngbuilds.io/.

robwormald committed:
ci: update chrome version to 67

robwormald committed:
ci: update firefox version to 61

robwormald committed:
test(core): update Web Platform feature detection

mary-poppins commented:
You can preview 8d290e2 at https://pr24861-8d290e2.ngbuilds.io/.

robwormald committed:
test(elements): add basic integration test for angular elements

mary-poppins commented:
You can preview 8cafe35 at https://pr24861-8cafe35.ngbuilds.io/.

mary-poppins commented:
You can preview 1f5de46 at https://pr24861-1f5de46.ngbuilds.io/.

robwormald committed:
feat(elements): enable Shadow DOM v1 and slots

When using ViewEncapsulation.ShadowDom, Angular will not remove the child nodes of the DOM node a root Component is bootstrapped into. This enables developers building Angular Elements to use the `<slot>` element to do native content projection.

mary-poppins commented:
You can preview f9b7adc at https://pr24861-f9b7adc.ngbuilds.io/.

gkalpak commented on packages/elements/test/slots_spec.ts:
```diff
@@ -0,0 +1,168 @@
+      const content = testContainer.querySelector('span.projected') !;
+      const slot = testEl.shadowRoot !.querySelector('slot') !;
+      const assignedNodes = slot.assignedNodes();
+      expect(assignedNodes[0]).toEqual(content);
```
This should be `toBe()`, instead of `toEqual()`, (here and below) to be sure the node is not duplicated or something.

gkalpak commented on packages/elements/test/slots_spec.ts:
```diff
@@ -0,0 +1,168 @@
+import {NgElement, NgElementConstructor, createCustomElement} from '../src/create-custom-element';
+import {NgElementStrategy, NgElementStrategyEvent, NgElementStrategyFactory} from '../src/element-strategy';
+
+type WithFooBar = {
```
Unused.

gkalpak commented on packages/elements/test/slots_spec.ts:
```diff
@@ -0,0 +1,168 @@
+      templateEl.innerHTML = tpl;
+      const template = templateEl.content.cloneNode(true) as DocumentFragment;
+      const testEl = template.querySelector('slot-events-el') !as NgElement & SlotEventsComponent;
+      const content = template.querySelector('span.projected');
```
Unused.

mary-poppins commented:
You can preview 41e0914 at https://pr24861-41e0914.ngbuilds.io/.

mhevery commented:
http://test/OCL:207128490:BASE:207128536:1533232686068:75ae6d95

IgorMinar commented on packages/core/src/render/api.ts:
```diff
@@ -267,7 +267,7 @@ export abstract class Renderer2 {
    * @returns The root element.
    */
-  abstract selectRootElement(selectorOrNode: string|any): any;
+  abstract selectRootElement(selectorOrNode: string|any, preserveContent?: boolean): any;
```
please document the new param

IgorMinar replied:
also please document the default value.

IgorMinar reviewed (changes requested):
this pr has many of unresolved review comments from @gkalpak please address.

robwormald commented on packages/platform-browser/testing/src/browser_util.ts:
```diff
@@ -68,6 +68,23 @@ export class BrowserDetection {
+
+  get supportsCustomElements() { return (typeof(<any>global).customElements !== 'undefined'); }
+
+  get supportsDeprecatedCustomCustomElementsV0() {
```
I want to add a number of new integration tests for native stuff, so I added it here, also to make clear that both of these APIs are DEPRECATED

mary-poppins commented:
You can preview 0f88a25 at https://pr24861-0f88a25.ngbuilds.io/.

mary-poppins commented:
You can preview ae10558 at https://pr24861-ae10558.ngbuilds.io/.

gkalpak commented on integration/ng_elements/e2e/app.e2e-spec.ts:
```diff
@@ -0,0 +1,20 @@
+      expect(helloWorldEl.getText()).toEqual('Hello World!');
+    });
+
+    it('should display: Hello fromIndex! via name attribute', function () {
```
_fromIndex_ --> _Foo_

gkalpak commented on integration/ng_elements/package.json:
```diff
@@ -0,0 +1,33 @@
+    "preprotractor": "tsc -p e2e",
+    "protractor": "protractor e2e/protractor.config.js"
+  }
+}
```
It's a good idea to have `private: true`, to avoid accidental publication.

gkalpak commented on integration/ng_elements/src/app.ts:
```diff
@@ -0,0 +1,33 @@
+    );
+    customElements.define(
+      'test-card',
+      createCustomElement(HelloWorldShadowComponent, {injector})
```
Use `TestCardComponent`.

gkalpak commented on integration/ng_elements/src/slots.html:
```diff
@@ -0,0 +1,20 @@
+
+</body>
+
+</html>
```
This file doesn't seem to be used atm :confused:

robwormald replied:
same as above - there's a whole suite of integration tests for slots/shadow roots I'd like to add, but can't until this PR lands :)

gkalpak commented on packages/core/src/render/api.ts:
```diff
@@ -265,9 +265,13 @@ export abstract class Renderer2 {
+   * @param preserveContent Whether the contents of the root element
+   * should be preserved, or cleared upon bootstrap (default behavior).
+   * Use with ViewEncapsulation.ShadowDom to allow simple native
+   * content projection via <slot> elements.
```
Wrap `ViewEncapsulation...` and `<slot>` in backticks to be rendered better in docs.

gkalpak reviewed (approved):
Some minor comments. LGTM otherwise :+1:

mary-poppins commented:
You can preview fe157a5 at https://pr24861-fe157a5.ngbuilds.io/.

robwormald commented:
final presubmit https://test.corp.google.com/ui#id=OCL:207948948:BASE:207949083:1533764780764:6e86095d

IgorMinar reviewed (approved):
a few more changes requested but otherwise it looks good.

please make one more additional change - we need better documentation of this change for the changelog. I'd argue that this PR is actually a feature that enables Angular Elements to use shadow dom and native projection via slots. Please update the relevant commit message to state that, otherwise this change will be lost in the sea of fixes and changes and developers will not know about it.

We should also add a small section to the existing elements guide that covers this feature. Please work with @jenniferfell on that. I think it's fine if that's done as a followup PR, with the intention that the blog post / release notes can point to a section in the guide where this feature is discussed.

mary-poppins commented:
You can preview e298282 at https://pr24861-e298282.ngbuilds.io/.

mary-poppins commented:
You can preview 687fc4e at https://pr24861-687fc4e.ngbuilds.io/.

IgorMinar commented:
@robwormald ci seems to be unhappy

mary-poppins commented:
You can preview 21bc468 at https://pr24861-21bc468.ngbuilds.io/.

robwormald committed:
fix(elements): strict null checks

mary-poppins commented:
You can preview ca3b8f1 at https://pr24861-ca3b8f1.ngbuilds.io/.

robwormald committed:
fix(elements): add compiler to integration

mary-poppins commented:
You can preview 32a0904 at https://pr24861-32a0904.ngbuilds.io/.

robwormald committed:
fix(elements): add compiler dependency

mary-poppins commented:
You can preview a81d30f at https://pr24861-a81d30f.ngbuilds.io/.
